plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php

```diff
@@ -0,0 +1,429 @@
+	 */
+	private function should_prevent_render() {
+		return ! $this->render_state['has_results'] && ! $this->render_state['has_no_results_block'];
+	}
+
+	/**
+	 * Reset the render state.
+	 */
+	private function reset_render_state() {
+		$this->render_state = array(
+			'has_results'          => false,
+			'has_no_results_block' => false,
+		);
+	}
+
+	/**
+	 * Enhances the Product Collection block with client-side pagination.
+	 *
+	 * This function identifies Product Collection blocks and adds necessary data attributes
+	 * to enable client-side navigation and animation effects. It also enqueues the Interactivity API runtime.
+	 *
+	 * @param string $block_content The HTML content of the block.
+	 * @param array  $block         Block details, including its attributes.
+	 *
+	 * @return string Updated block content with added interactivity attributes.
+	 */
+	public function enhance_product_collection_with_interactivity( $block_content, $block ) {
+		$is_product_collection_block = $block['attrs']['query']['isProductCollectionBlock'] ?? false;
+
+		if ( $is_product_collection_block ) {
+			// Enqueue the Interactivity API runtime and set the namespace.
+			wp_enqueue_script( 'wc-interactivity' );
+			$p = new \WP_HTML_Tag_Processor( $block_content );
+			if ( $this->is_next_tag_product_collection( $p ) ) {
+				$this->set_product_collection_namespace( $p );
+			}
+			// Check if dimensions need to be set and handle accordingly.
+			$this->handle_block_dimensions( $p, $block );
+			$block_content = $p->get_updated_html();
+
+			$collection    = $block['attrs']['collection'] ?? '';
+			$block_content = $this->add_rendering_callback( $block_content, $collection );
+
+			$is_enhanced_pagination_enabled = ! ( $block['attrs']['forcePageReload'] ?? false );
+			if ( $is_enhanced_pagination_enabled ) {
+				$block_content = $this->enable_client_side_navigation( $block_content );
+			}
+		}
+		return $block_content;
+	}
+
+	/**
+	 * Check if next tag is a PC block.
+	 *
+	 * @param WP_HTML_Tag_processor $p Initial tag processor.
+	 *
+	 * @return bool Answer if PC block is available.
+	 */
+	private function is_next_tag_product_collection( $p ) {
+		return $p->next_tag( array( 'class_name' => 'wp-block-woocommerce-product-collection' ) );
+	}
+
+	/**
+	 * Set PC block namespace for Interactivity API.
+	 *
+	 * @param WP_HTML_Tag_processor $p Initial tag processor.
+	 */
+	private function set_product_collection_namespace( $p ) {
+		$p->set_attribute( 'data-wc-interactive', wp_json_encode( array( 'namespace' => 'woocommerce/product-collection' ), JSON_HEX_TAG | JSON_HEX_APOS | JSON_HEX_QUOT | JSON_HEX_AMP ) );
+	}
+
+	/**
+	 * Set the style attribute for fixed width.
+	 *
+	 * @param WP_HTML_Tag_Processor $p          The HTML tag processor.
+	 * @param string                $fixed_width The fixed width value.
+	 */
+	private function set_fixed_width_style( $p, $fixed_width ) {
+		$p->set_attribute( 'style', $this->get_list_styles( $fixed_width ) );
+	}
+
```

Comment:
![image](https://github.com/user-attachments/assets/7e1a2f64-1f01-44de-a1cb-9d5465cc3dad)

Reply:
Method moved in the commit [de1dbe8](https://github.com/woocommerce/woocommerce/pull/52429/commits/de1dbe8c6e1e9b014bab398d1c2dc3c96e2b29b0)